sympy/tensor/functions.py

```diff
@@ -61,3 +63,46 @@ def __getitem__(self, index):
             arg.__getitem__(tuple(next(index) for i in shp))
             for arg, shp in zip(self.args, self._get_args_shapes())
         )
+
+
+def shape(expr):
+    """
+    Return the shape of the tensor.
+
+    This function dispatches over the kind and class of *expr* to return
+    its shape.
```

Comment:
If `shape` is an end user function then this should not be mentioned at the top of the docstring.

Perhaps in a "Notes" section at the bottom there could be a brief mention that `accept_shape` is used as the mechanism for implementing this function and can be used to extend it to define the shape of other kinds of objects so that `shape()`. There can be a "See also" referencing `accept_shape` and then the `accept_shape` docstring can explain how to overload shape for a new kind and how to implement e.g. matrix shape for a new class.

Reply:
What is worth mentioning is the contrast between `shape(obj)` and `obj.shape` and the fact that `shape(obj)` applies in a more general context. We should make sure that there are "See also" referencing in both directions from `.shape` and `shape()`.

Probably the matrices tutorial should be updated to use `shape()` or should at least mention it.